jax/interpreters/numpy_eval.py

```diff
@@ -0,0 +1,767 @@
+    np.maximum(0, np.multiply(interior, np.subtract(operand.shape, 1))))
+  out = np.full(outshape, padding_value, operand.dtype)
+  lhs_slices = tuple(_slice(l if l > 0 else 0, -h if h > 0 else None, step)
+                     for l, h, step in zip(lo_pos, hi_pos, np.add(1, interior)))
+  out[lhs_slices] = operand
+  trim_slices = tuple(_slice(-l if l < 0 else 0, h if h < 0 else None)
+                      for l, h in zip(lo, hi))
+  return out[trim_slices]
+np_impl[lax.pad_p] = pad
+
+def rev(operand, dimensions):
+  dimensions = frozenset(dimensions)
+  indexer = (_slice(None, None, -1) if d in dimensions else _slice(None)
+             for d in range(np.ndim(operand)))
+  return operand[tuple(indexer)]
+np_impl[lax.rev_p] = rev
+
+select = np.where
+np_impl[lax.select_p] = select
+
+def slice(operand, start_indices, limit_indices, strides=None):  # pylint: disable=redefined-builtin
+  if strides is None:
+    strides = np.ones(len(start_indices)).astype(int)
+  idx = tuple(map(_slice, start_indices, limit_indices, strides))
+  return _ensure_numpy(operand)[idx]
+np_impl[lax.slice_p] = slice
+
+def dynamic_slice(operand, *start_indices, slice_sizes):
+  out = np.zeros(slice_sizes, dtype=operand.dtype)
+  idx = tuple(_slice(start, start+size)
+              for start, size in zip(start_indices, slice_sizes))
+  section = _ensure_numpy(operand)[idx]
+  out[tuple(_slice(None, stop) for stop in section.shape)] = section
+  return out
+np_impl[lax.dynamic_slice_p] = dynamic_slice
+
+def dynamic_update_slice(operand, update, *start_indices):
+  slices = tuple(map(_slice, start_indices, np.add(start_indices, update.shape)))
+  updated_operand = np.copy(operand)
+  updated_operand[slices] = update
+  return updated_operand
+np_impl[lax.dynamic_update_slice_p] = dynamic_update_slice
+
+def transpose(operand: Array, permutation: Sequence[int]) -> Array:
+  return np.transpose(_ensure_numpy(operand), permutation)
+np_impl[lax.transpose_p] = transpose
+
+@curry
+def _reduce(reducer: np.ufunc, operand: Array, axes: Sequence[int]) -> np.ndarray:
+  operand = np.asarray(operand)
+  dtype = operand.dtype
+  if dtype == dtypes.bfloat16:
+    operand = operand.astype(np.float64)
+  elif dtype == np.uint64 and not config.FLAGS.jax_enable_x64:
+    operand = operand.astype(np.uint32)
+  out = reducer.reduce(operand, axis=tuple(axes))
+  if dtype == dtypes.bfloat16:
+    out = out.astype(np.float64)
+  return out.astype(dtype)
+np_impl[lax.reduce_sum_p] = _reduce(np.add)
+np_impl[lax.reduce_prod_p] = _reduce(np.multiply)
+np_impl[lax.reduce_max_p] = _reduce(np.maximum)
+np_impl[lax.reduce_min_p] = _reduce(np.minimum)
+np_impl[lax.reduce_or_p] = _reduce(np.logical_or)
+np_impl[lax.reduce_and_p] = _reduce(np.logical_and)
+
+def reduce(operand: Array, init_value: Array, computation: Callable,
+           dimensions: Sequence[int], jaxpr=None, consts=None) -> np.ndarray:
+  reducer = _reducer(computation, init_value)
```

Comment:
Brought up in #4109: I'd recommend ignoring `computation` here, and using `core.jaxpr_as_fun` on `jaxpr` instead to reliably obtain a callable form of the reduction binop.

Reply:
@froystig Done.